dot/parachain/spam_slots.go

```diff
@@ -0,0 +1,200 @@
+func (s *spamSlots) getValidators(session parachain.SessionIndex, candidate common.Hash) (*treeset.Set, bool) {
+	s.unconfirmedLock.RLock()
+	defer s.unconfirmedLock.RUnlock()
+	validators, ok := s.unconfirmed[unconfirmedDisputesKey(session, candidate)]
+	return validators, ok
+}
+
+func (s *spamSlots) clearValidators(session parachain.SessionIndex, candidate common.Hash) {
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+	delete(s.unconfirmed, unconfirmedDisputesKey(session, candidate))
+}
+
+func (s *spamSlots) clearSlots(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	delete(s.slots, slotsKey(session, validator))
+}
+
+func (s *spamSlots) incrementSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	s.slots[slotsKey(session, validator)]++
+}
+
+func (s *spamSlots) AddUnconfirmed(session parachain.SessionIndex, candidate common.Hash, validator parachain.ValidatorIndex) bool {
+	if spamCount, _ := s.getSpamCount(session, validator); spamCount >= s.maxSpamVotes {
+		return false
+	}
+
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+
+	validators, ok := s.unconfirmed[unconfirmedDisputesKey(session, candidate)]
+	if !ok || validators == nil {
+		validators = treeset.NewWith(byValidatorIndex)
+	}
+
+	if !validators.Contains(validator) {
+		validators.Add(validator)
+		s.incrementSpamCount(session, validator)
+	}
+
+	s.unconfirmed[unconfirmedDisputesKey(session, candidate)] = validators
+
+	return true
+}
+
+func (s *spamSlots) Clear(session parachain.SessionIndex, candidate common.Hash) {
+	if validators, ok := s.getValidators(session, candidate); ok {
+		validatorSet := validators.Values()
+		s.clearValidators(session, candidate)
+
+		for _, validator := range validatorSet {
+			spamCount, ok := s.getSpamCount(session, validator.(parachain.ValidatorIndex))
+			if ok {
+				if spamCount == 1 {
+					s.clearSlots(session, validator.(parachain.ValidatorIndex))
+					continue
+				}
+
+				s.slotsLock.Lock()
+				s.slots[slotsKey(session, validator.(parachain.ValidatorIndex))] = spamCount - 1
+				s.slotsLock.Unlock()
+			}
+		}
+	}
+}
+
+func (s *spamSlots) PruneOld(oldestIndex parachain.SessionIndex) {
+	s.unconfirmedLock.Lock()
+	unconfirmedToDelete := make([]unconfirmedKey, 0)
+	for k := range s.unconfirmed {
+		if k.session < oldestIndex {
+			unconfirmedToDelete = append(unconfirmedToDelete, k)
+		}
+	}
+
+	for _, k := range unconfirmedToDelete {
+		delete(s.unconfirmed, k)
+	}
+	s.unconfirmedLock.Unlock()
+
+	s.slotsLock.Lock()
+	slotsToDelete := make([]slotKey, 0)
+	for k := range s.slots {
+		if k.session < oldestIndex {
+			slotsToDelete = append(slotsToDelete, k)
+		}
+	}
+
+	for _, k := range slotsToDelete {
+		delete(s.slots, k)
+	}
+	s.slotsLock.Unlock()
+}
+
+// NewSpamSlots returns a new SpamSlots instance
+func NewSpamSlots(maxSpamVotes uint32) SpamSlots {
```

Comment:
I would suggest returning the concrete type instead of the interface, and since Golang supports implicit interface you can still return a private struct that implements the interface methods and let the caller to receive the interface 

```go
// dot/parachain pkg
func NewSpamSlots(maxSpamVotes uint32) *spamSlots { ... }

// dot/sync pkg
func startSync() {
    sync(parachain.NewSpamSlots(10)) /* <- return the concrete type which implements the expected interface*/)
}
func sync(spamSlots parachain.SpamSlots /* <- interface */) {}

```

Reply:
Hmm, the reason why I return an interface is to make sure that the struct implements all the interface methods. If it doesn't, there will be a compile time error at `NewSpamSlots()`. 

If we return just the struct, I guess it will throw an error at `sync(parachain.NewSpamSlots(10))`. I'd rather handle this at the calling package rather than at the caller.

Is there like a convention that says we shouldn't return an interface?

Reply:
> Is there like a convention that says we shouldn't return an interface?

Mostly what I read is `accept interfaces return structs` but this is not a rule, the point is the consumer should be free to receive what interface it wants, for example, every time you want to introduce a new behavior for a certain struct you will need to update the interface too and all the consumers will be affected, 

Refs:
- https://mycodesmells.com/post/accept-interfaces-return-struct-in-go
- https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b

> Hmm, the reason why I return an interface is to make sure that the struct implements all the interface methods. If it doesn't, there will be a compile-time error at NewSpamSlots().

In some parts of the code, you will see the following pattern to ensure the implementation at comptime

```go
// ensure at compile time the struct implements the interface
var _ Message = &RequestBlockMessage{} 

type Message interface {
    Encode() ([]byte, error)
}

type RequestBlockMessage struct {...}
func (r *RequestBlockMessage) Encode() ([]byte, error) {...}
```

> If we return just the struct, I guess it will throw an error at sync(parachain.NewSpamSlots(10)). I'd rather handle this at the calling package than at the caller.

No, there is no error. We expect a public interface as an argument and the returned struct to implement those methods

Reply:
> No, there is no error. We expect a public interface as an argument and the returned struct to implement those methods

I meant if the struct doesn't implement all the methods of an interface, it would throw an error at `sync(parachain.NewSpamSlots(10))`

There is always something new to learn everyday! Thank you for sharing @EclesioMeloJunior